File: lessons/04_03/Java/1_base_classes.java
class Animal {
  private int health;
  private float speed;
  private String environment;

  public Animal(int health, float speed, String environment) {
    this.health = health;
    this.speed = speed;
    this.environment = environment;
  }

  public int getHealth() {
    return health;
  }

  public void setHealth(int health) {
    this.health = health;
  }

  public float getSpeed() {
    return speed;
  }

  public void setSpeed(float speed) {
    this.speed = speed;
  }

  public void animalInfo() {
    System.out.println("Animal lives in " + environment);
  }
}

class Pigeon extends Animal {
  public Pigeon(int health, float speed, String environment) {
    super(health, speed, environment);
  }

  public static void titlePrint() {
    System.out.println("\tPIGEON INFO:");
  }
}

class Dog extends Animal {
  public Dog(int health, float speed, String environment) {
    super(health, speed, environment);
  }

  public static void titlePrint() {
    System.out.println("\tDOG INFO:");
  }
}

class Dolphin extends Animal {
  public Dolphin(int health, float speed, String environment) {
    super(health, speed, environment);
  }

  public static void titlePrint() {
    System.out.println("\tDOLPHIN INFO:");
  }
}

public class Main {
  public static void main(String[] args) {
    Pigeon pigeon = new Pigeon(100, 40, "Air");
    Dog dog = new Dog(90, 25, "Earth");
    Dolphin dolphin = new Dolphin(65, 70, "Water");

    Pigeon.titlePrint();
    System.out.println("Pigeon health: " + pigeon.getHealth());
    System.out.println("Pigeon speed: " + pigeon.getSpeed());
    pigeon.animalInfo();
    System.out.println();

    Dog.titlePrint();
    System.out.println("Dog health: " + dog.getHealth());
    System.out.println("Dog speed: " + dog.getSpeed());
    dog.animalInfo();
    System.out.println();

    Dolphin.titlePrint();
    System.out.println("Dolphin health: " + dolphin.getHealth());
    System.out.println("Dolphin speed: " + dolphin.getSpeed());
    dolphin.animalInfo();
    System.out.println();
  }
}

File: lessons/04_03/Golang/1_base_classes.go
// В Golang нет ООП как во всех ЯП
// В Go есть структруы и возможность задавать определенные функции определенным структурам
// Наследование реализуется путем вложения структур друг в друга

package main

import "fmt"

type Animal struct {
	health      int
	speed       float64
	environment string
}

func (a *Animal) GetHealth() int {
	return a.health
}

func (a *Animal) SetHealth(health int) {
	a.health = health
}

func (a *Animal) GetSpeed() float64 {
	return a.speed
}

func (a *Animal) SetSpeed(speed float64) {
	a.speed = speed
}

func (a *Animal) AnimalInfo() {
	fmt.Println("Animal lives in", a.environment)
}

type Pigeon struct {
	Animal
}

func NewPigeon(health int, speed float64, environment string) *Pigeon {
	return &Pigeon{Animal{health, speed, environment}}
}

func (p *Pigeon) TitlePrint() {
	fmt.Println("\tPIGEON INFO:")
}

type Dog struct {
	Animal
}

func NewDog(health int, speed float64, environment string) *Dog {
	return &Dog{Animal{health, speed, environment}}
}

func (d *Dog) TitlePrint() {
	fmt.Println("\tDOG INFO:")
}

type Dolphin struct {
	Animal
}

func NewDolphin(health int, speed float64, environment string) *Dolphin {
	return &Dolphin{Animal{health, speed, environment}}
}

func (d *Dolphin) TitlePrint() {
	fmt.Println("\tDOLPHIN INFO:")
}

func main() {
	pigeon := NewPigeon(100, 40, "Air")
	dog := NewDog(90, 25, "Earth")
	dolphin := NewDolphin(65, 70, "Water")

	pigeon.TitlePrint()
	fmt.Println("Pigeon health:", pigeon.GetHealth())
	fmt.Println("Pigeon speed:", pigeon.GetSpeed())
	pigeon.AnimalInfo()
	fmt.Println()

	dog.TitlePrint()
	fmt.Println("Dog health:", dog.GetHealth())
	fmt.Println("Dog speed:", dog.GetSpeed())
	dog.AnimalInfo()
	fmt.Println()

	dolphin.TitlePrint()
	fmt.Println("Dolphin health:", dolphin.GetHealth())
	fmt.Println("Dolphin speed:", dolphin.GetSpeed())
	dolphin.AnimalInfo()
	fmt.Println()
}
